pandas/_libs/lib.pyx

```diff
@@ -721,16 +735,27 @@ cpdef ndarray[object] ensure_string_array(
     if copy and result is arr:
         result = result.copy()
 
+    if coerce == 'strict-null':
+        # We don't use checknull, since NaT, Decimal("NaN"), etc. aren't valid
+        # If they are present, they are treated like a regular Python object
+        # and will either cause an exception to be raised or be coerced.
+        check_null = strict_na_values.__contains__
+    else:
+        check_null = checknull
+
     for i in range(n):
         val = arr[i]
 
         if isinstance(val, str):
             continue
 
-        if not checknull(val):
-            result[i] = str(val)
+        if not check_null(val):
+            if coerce =="all" or coerce == "non-null":
```

Comment:
I'm kind of praying that the C compiler/Cython will optimize this for us, since moving this outside would cause a ton of code duplication. The string asvs didn't show any perf changes(but there's a lot of n/a's when running benchmarks) so idk.

Reply:
Update: ASV's (see bottom comment) are unaffected so gonna assume the compiler hoisted the conditions.